runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ArtifactServiceStager.java

```diff
@@ -87,26 +87,32 @@ private ArtifactServiceStager(Channel channel, int bufferSize) {
     this.bufferSize = bufferSize;
   }
 
-  public void stage(Iterable<File> files) throws IOException, InterruptedException {
-    final Map<File, CompletionStage<ArtifactMetadata>> futures = new HashMap<>();
-    for (File file : files) {
+  /**
+   * Stages the given artifact files to the staging service.
+   *
+   * @return The artifact staging token returned by the service
+   */
+  public String stage(Iterable<StagedFile> files) throws IOException, InterruptedException {
+    final Map<StagedFile, CompletionStage<ArtifactMetadata>> futures = new HashMap<>();
+    for (StagedFile file : files) {
       futures.put(file, MoreFutures.supplyAsync(new StagingCallable(file), executorService));
     }
     CompletionStage<StagingResult> stagingResult =
         MoreFutures.allAsList(futures.values())
             .thenApply(ignored -> new ExtractStagingResultsCallable(futures).call());
-    stageManifest(stagingResult);
+    return stageManifest(stagingResult);
   }
 
-  private void stageManifest(CompletionStage<StagingResult> stagingFuture)
+  private String stageManifest(CompletionStage<StagingResult> stagingFuture)
       throws InterruptedException {
     try {
       StagingResult stagingResult = MoreFutures.get(stagingFuture);
       if (stagingResult.isSuccess()) {
         Manifest manifest =
             Manifest.newBuilder().addAllArtifact(stagingResult.getMetadata()).build();
-        blockingStub.commitManifest(
-            CommitManifestRequest.newBuilder().setManifest(manifest).build());
+        return blockingStub
+            .commitManifest(CommitManifestRequest.newBuilder().setManifest(manifest).build())
+            .getStagingToken();
```

Comment:
Separate these two lines? Just a preference thing, but I think that the fluent style hints at something more immediate than the actual RPC we're making here.